File: ts/src/connection-types.ts
import { u32 } from "./seq.js";
import { FlagSet, Flags, Segment } from "./wire.js";

export interface TrackedSegment {
  seq: number;
  flags: FlagSet;
  payload: Uint8Array;
  sentAtMs: number;
  retransmitted: boolean;
  transmissions: number;
}

export interface ReassemblySegment {
  seq: number;
  payload: Uint8Array;
  fin: boolean;
}

export interface ConnectionUpdate {
  segments: Segment[];
  accepted: boolean;
  closed: boolean;
}

export interface AckOutcome {
  finAcked: boolean;
  retransmit?: Segment;
}

export const openUpdate = (segments: Segment[] = []): ConnectionUpdate => ({
  segments,
  accepted: false,
  closed: false,
});

export const trackedEnd = (segment: TrackedSegment): number =>
  u32(
    segment.seq +
      segment.payload.length +
      Number(segment.flags.has(Flags.Syn)) +
      Number(segment.flags.has(Flags.Fin)),
  );

export const reassemblyEnd = (segment: ReassemblySegment): number =>
  u32(segment.seq + segment.payload.length + Number(segment.fin));

File: rust/fips-tcp/src/connection_types.rs
use crate::wire::Flags;

#[derive(Clone, Debug)]
pub(crate) struct TrackedSegment {
    pub(crate) seq: u32,
    pub(crate) flags: Flags,
    pub(crate) payload: Vec<u8>,
    pub(crate) sent_at_ms: u64,
    pub(crate) retransmitted: bool,
    pub(crate) transmissions: u8,
}

impl TrackedSegment {
    pub(crate) fn end_seq(&self) -> u32 {
        self.seq
            .wrapping_add(self.payload.len() as u32)
            .wrapping_add(u32::from(self.flags.contains(Flags::SYN)))
            .wrapping_add(u32::from(self.flags.contains(Flags::FIN)))
    }
}

#[derive(Clone, Debug)]
pub(crate) struct ReassemblySegment {
    pub(crate) seq: u32,
    pub(crate) payload: Vec<u8>,
    pub(crate) fin: bool,
}

impl ReassemblySegment {
    pub(crate) fn end_seq(&self) -> u32 {
        self.seq
            .wrapping_add(self.payload.len() as u32)
            .wrapping_add(u32::from(self.fin))
    }
}
